llvm/lib/Target/NVPTX/SYCL/GlobalOffset.cpp

```diff
@@ -0,0 +1,368 @@
+          ImplicitOffsets.push_back(U);
+        }
+        ImplicitOffsets.push_back(ImplicitOffset);
+
+        // Replace call to other function (which now has a new parameter),
+        // with a call including the new parameter to that same function.
+        auto NewCaller = CallInst::Create(
+            /* Ty= */ CalleeWithImplicitParam->getFunctionType(),
+            /* Func= */ CalleeWithImplicitParam,
+            /* Args= */ ImplicitOffsets,
+            /* NameStr= */ Twine(),
+            /* InsertBefore= */ CallToOld);
+        NewCaller->setTailCallKind(CallToOld->getTailCallKind());
+        CallToOld->replaceAllUsesWith(NewCaller);
+
+        if (CallToOld->hasName()) {
+          NewCaller->takeName(CallToOld);
+        }
+      }
+
+      // Remove the caller now that it has been replaced.
+      CallToOld->eraseFromParent();
+
+      if (!AlreadyProcessed) {
+        // Process callers of the old function.
+        this->addImplicitParameterToCallers(M, Caller, NewFunc);
+
+        // Now that the old function is dead, delete it.
+        Caller->dropAllReferences();
+        Caller->eraseFromParent();
+      }
+    }
+  }
+
+  std::pair<Function *, Value *>
+  addOffsetArgumentToFunction(Module &M, Function *Func,
+                              Type *ImplicitArgumentType = nullptr,
+                              bool KeepOriginal = false) {
+    FunctionType *FuncTy = Func->getFunctionType();
+    const AttributeList &FuncAttrs = Func->getAttributes();
+    ImplicitArgumentType =
+        ImplicitArgumentType ? ImplicitArgumentType : ImplicitOffsetPtrType;
+
+    // Construct an argument list containing all of the previous arguments.
+    SmallVector<Type *, 8> Arguments;
+    SmallVector<AttributeSet, 8> ArgumentAttributes;
+
+    unsigned i = 0;
+    for (Function::arg_iterator FuncArg = Func->arg_begin(),
+                                FuncEnd = Func->arg_end();
+         FuncArg != FuncEnd; ++FuncArg, ++i) {
+      Arguments.push_back(FuncArg->getType());
+      ArgumentAttributes.push_back(FuncAttrs.getParamAttributes(i));
+    }
+
+    // Add the offset argument. Must be the same type as returned by
+    // `llvm.nvvm.implicit.offset`.
+
+    Arguments.push_back(ImplicitArgumentType);
+    ArgumentAttributes.push_back(AttributeSet());
+
+    // Build the new function.
+    AttributeList NAttrs =
+        AttributeList::get(Func->getContext(), FuncAttrs.getFnAttributes(),
+                           FuncAttrs.getRetAttributes(), ArgumentAttributes);
+    assert(!FuncTy->isVarArg() && "Variadic arguments prohibited in SYCL");
+    FunctionType *NewFuncTy = FunctionType::get(FuncTy->getReturnType(),
+                                                Arguments, FuncTy->isVarArg());
+
+    Function *NewFunc = Function::Create(NewFuncTy, Func->getLinkage(),
+                                         Func->getAddressSpace());
+
+    if (KeepOriginal) {
+      NewFunc->setName(Func->getName() + "_with_offset");
```

Comment:
NOTE: this change breaks original mangling, so if other tools rely on mangling rules to unmangle the name of the function (e.g. profiler, debugger, etc.) they might have problems with decoding this suffix. I think we should add a TODO comment to address this later.

Reply:
I've added a TODO comment, but I wonder how to make sure that the PI CUDA backend would be able to easily find the new kernels if the naming changes.

Reply:
The best way is to not add/rename functions.
What would be the implications of always having offset parameter?

Reply:
Having the offset parameter on all entry points would increase the number of registers used by each thread unnecessarily in most cases. In this implementation the compiler is able to optimize out the 0-offset inserted into the original function.

Reply:
So when a kernel is called and the offset is 0, there is no performance impact?
How does it work? With static analysis or compiling 2 versions and picking the optimized one at runtime?
Does anyone has some statistics about the utilization of global offset? I have never used it...

Reply:
> How does it work? With static analysis or compiling 2 versions and picking the optimized one at runtime?

The latter. It clones most[1] kernel and adds an offset parameter to the clone, then it inserts an array of three 0 into the original kernel, which is then used in all places where either the global offset intrinsic is used or any function that - either directly or down the line - may use the global offset intrinsic. Luckily the compiler is smart enough to figure out that all uses of elements in this array - which remains unchanged throughout - can be simplified and thus the array will in turn be removed at some point during optimization. The PI CUDA backend then selects the version of the kernel with the offset parameter iff any element of the given offset is strictly greater than 0, otherwise it selects the "original".

[1] Kernels that is completely invariant w.r.t. global offset is not cloned or altered.

> Does anyone has some statistics about the utilization of global offset? I have never used it...

Me neither, but I have seen it being used a couple of times. I can see it being useful.